### ISSUE TITLE
Support per-project YAML config files (search current directory)

### ISSUE THREAD

**Issue by gjoseph92** [dask/dask#9702]
Dask currently supports YAML config files at a number of _global_ paths: https://docs.dask.org/en/stable/configuration.html#yaml-files. However, it can be useful to set configuration per-project.

You can currently do this by creating a YAML file in your repo (say `dask.yaml`) and setting `DASK_CONFIG=dask.yaml`. But then you need to remember to set the environment variable all the time, or use dotenv or something.

It would be nice if Dask would also also search for a `dask.yaml` file (name up for debate) in the current directory, and have it take precedence over the other layers of config. Possibly it could also traverse up from the current directory until it found a `dask.yaml` file.

**Comment by quasiben** [comment]
I would be +1 on that functionality. Last month we chatted a bit about trying to improve how configuration location could potentially be improved: https://github.com/rapidsai/dask-cuda/issues/1020

**Comment by gjoseph92** [comment]
@quasiben is this something your team might be interested in working on?

**Comment by quasiben** [comment]
Yeah, I think we could probably take this on.  Let's use this thread to iterate on design / get broad agreement.  Right now the proposal is just search locally -- do you think we'd want some mechanism for the user to ask where the configuration is being loaded from ?  For example, conda can do this in the following way:

```
(base) bzaitlen@dgx15:~$ conda config --show-sources
==> /home/nfs/bzaitlen/.condarc <==
channels:
  - conda-forge
  - defaults
report_errors: False
```

We also explored providing this feedback in dask-cuda with https://github.com/rapidsai/dask-cuda/pull/1006 (though the name is not great and may also change) and asked about it here: https://github.com/dask/distributed/issues/7106

My guess is this shouldn't be too hard to implement but we should just check with others like @jacobtomlinson / @jrbourbeau to make sure this isn't too "magical"

**Comment by gjoseph92** [comment]
> do you think we'd want some mechanism for the user to ask where the configuration is being loaded from

That sounds handy, but I'd hope it could also be a separate task.

> Right now the proposal is just search locally

Yeah, I'm thinking:
1. minimal: if there's a `dask.yaml` or `dask.yml` file in the current directory, use it (on top of other configs)
2. nicer?: walk from the current directory up to `/`, collecting files named `dask.yaml` or `dask.yml`. Layer them on top of other configs.

**Comment by jrbourbeau** [comment]
Apologies for the delayed reply. My initial reaction was "our config system already has lots of places it looks, do we really need another?" But after thinking about it more, I'm +0.5 for adding support for a config file in the current working directly. 

Things I don't like about it:

- It's yet another place where config files could be

Things I do like about it:

- Makes it easier to include/use a Dask config file in a project (e.g. `git add .dask.yaml` + `git commit -m "Adds custom Dask config"`)
    - @quasiben mentioned offline that he's run into multiple user groups who wanted an easy way to have a different config file per project. The change proposed here would resolve that user pain. 
- Easier to remember for Dask devs who want to experiment locally (this is @gjoseph92 wants) 
- This change is backwards compatible.

Also cc @djhoese for visibility

**Comment by djhoese** [comment]
I'm also +0.5. From what I've seen of recent dask and xarray work, people are trying to get import time down. This is another iteration over something on disk that adds to import time. And as someone who has projects that allow configuration files/directories in the PWD, it becomes yet another thing you always have to ask users on support tickets (ex. "Any chance you have a file named `dask.yaml` in your working directory?").

I agree that it would be nice for projects and developers to quickly customize their configuration and for a single use case scenario that maybe makes sense, but I'm not so sure for project-based use cases. Maybe I'm missing something. What kind of dask configuration options are people setting on a per-project basis? Is this like an ML project that knows that the settings for fusing or other optimizations should be set to X because they product the best performance for their workflow? Or is this a infrastructure as code type of situation where a user just wants to specify the dask configuration for this deployment?

Lastly, how often is a user's custom configuration limited to a single config file? Dask configs don't support the idea of "profiles" right? If I'm a developer switching between two or more configuration cases while debugging, then detecting a single config file in my local directory isn't the perfect solution (as I'd need to rename the files to get it to auto-detect the configuration I'm testing). But maybe this isn't a use case for this feature. But I wanted to bring it up in case there is a way to support both use cases.

`</brainstorm>`
